-1

**Failed Tests**:   Build 
**Size**: This PR adds an extra 64KB to repository
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-bfad9e/44235/summary.html
**COMMIT**: d5d136be1da3032109be9b4d4749b0b1bc5fdc8b
**CMSSW**: CMSSW_15_0_X_2025-02-05-2300/el8_amd64_gcc12
**Additional Tests**: GPU
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/47280/44235/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Compiling  src/HeterogeneousCore/ROCmServices/test/testROCmService.cpp
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -D__HIP_PLATFORM_HCC__ -D__HIP_PLATFORM_AMD__ -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH='CMSSW_15_0_X_2025-02-05-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_15_0_X_2025-02-05-2300' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_15_0_X_2025-02-05-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/pcre/8.43-2d141998cfe5424b8f7aff48035cc2da/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/boost/1.80.0-a2c84315bd72151dcb3b6e3fe5018437/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/python3/3.9.14-ccc34bac15aa449b4c76ba24d02d2fd7/include/python3.9 -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/rocm/6.2.4-0a366585c16e3116ddeaba7741c05c93/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/lcg/root/6.32.09-47cefdd6f737afcf5535fa577c928c47/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/tbb/v2021.9.0-63e1493f6c63f7899f38cf6d13a1d19f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/zlib/1.2.13-d217cdbdd8d586e845e05946de2796be/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/catch2/2.13.6-17102db92de47c6a473c6e67627c548a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/fmt/10.2.1-e35fd1db5eb3abc8ac0452e8ee427196/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/py3-pybind11/2.13.6-16793d3657f4e0749f8f9007c2eabf31/lib/python3.9/site-packages/pybind11/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/tinyxml2/6.2.0-f99ae2781d074227d47e8a3e7c8ec87e/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/HeterogeneousCore/ROCmServices/test/testROCmService/testROCmService.cpp.d src/HeterogeneousCore/ROCmServices/test/testROCmService.cpp -o tmp/el8_amd64_gcc12/src/HeterogeneousCore/ROCmServices/test/testROCmService/testROCmService.cpp.o
>> Compiling  src/HeterogeneousCore/ROCmServices/test/test_main.cpp
/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/gcc/12.3.1-40d504be6370b5a30e3947a6e575ca28/bin/c++ -c -DCMS_MICRO_ARCH='x86-64-v3' -DGNU_GCC -D_GNU_SOURCE -DTBB_USE_GLIBCXX_VERSION=120301 -DTBB_SUPPRESS_DEPRECATED_MESSAGES -DTBB_PREVIEW_RESUMABLE_TASKS=1 -DTBB_PREVIEW_TASK_GROUP_EXTENSIONS=1 -D__HIP_PLATFORM_HCC__ -D__HIP_PLATFORM_AMD__ -DBOOST_SPIRIT_THREADSAFE -DPHOENIX_THREADSAFE -DBOOST_MATH_DISABLE_STD_FPCLASSIFY -DBOOST_UUID_RANDOM_PROVIDER_FORCE_POSIX -DCMSSW_GIT_HASH='CMSSW_15_0_X_2025-02-05-2300' -DPROJECT_NAME='CMSSW' -DPROJECT_VERSION='CMSSW_15_0_X_2025-02-05-2300' -Isrc -Ipoison -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/cms/cmssw-patch/CMSSW_15_0_X_2025-02-05-2300/src -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/pcre/8.43-2d141998cfe5424b8f7aff48035cc2da/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/boost/1.80.0-a2c84315bd72151dcb3b6e3fe5018437/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/bz2lib/1.0.6-d065ccd79984efc6d4660f410e4c81de/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/libuuid/2.34-27ce4c3579b5b1de2808ea9c4cd8ed29/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/python3/3.9.14-ccc34bac15aa449b4c76ba24d02d2fd7/include/python3.9 -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/rocm/6.2.4-0a366585c16e3116ddeaba7741c05c93/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/lcg/root/6.32.09-47cefdd6f737afcf5535fa577c928c47/include -isystem/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/tbb/v2021.9.0-63e1493f6c63f7899f38cf6d13a1d19f/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/xz/5.2.5-6f3f49b07db84e10c9be594a1176c114/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/zlib/1.2.13-d217cdbdd8d586e845e05946de2796be/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/catch2/2.13.6-17102db92de47c6a473c6e67627c548a/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/fmt/10.2.1-e35fd1db5eb3abc8ac0452e8ee427196/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/md5/1.0.0-5b594b264e04ae51e893b1d69a797ec6/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/py3-pybind11/2.13.6-16793d3657f4e0749f8f9007c2eabf31/lib/python3.9/site-packages/pybind11/include -I/cvmfs/cms-ib.cern.ch/sw/x86_64/nweek-02875/el8_amd64_gcc12/external/tinyxml2/6.2.0-f99ae2781d074227d47e8a3e7c8ec87e/include -O3 -pthread -pipe -Werror=main -Werror=pointer-arith -Werror=overlength-strings -Wno-vla -Werror=overflow -std=c++20 -ftree-vectorize -Werror=array-bounds -Werror=format-contains-nul -Werror=type-limits -fvisibility-inlines-hidden -fno-math-errno --param vect-max-version-for-alias-checks=50 -Xassembler --compress-debug-sections -Wno-error=array-bounds -Warray-bounds -fuse-ld=bfd -march=x86-64-v3 -felide-constructors -fmessage-length=0 -Wall -Wno-non-template-friend -Wno-long-long -Wreturn-type -Wextra -Wpessimizing-move -Wclass-memaccess -Wno-cast-function-type -Wno-unused-but-set-parameter -Wno-ignored-qualifiers -Wno-unused-parameter -Wunused -Wparentheses -Werror=return-type -Werror=missing-braces -Werror=unused-value -Werror=unused-label -Werror=address -Werror=format -Werror=sign-compare -Werror=write-strings -Werror=delete-non-virtual-dtor -Werror=strict-aliasing -Werror=narrowing -Werror=unused-but-set-variable -Werror=reorder -Werror=unused-variable -Werror=conversion-null -Werror=return-local-addr -Wnon-virtual-dtor -Werror=switch -fdiagnostics-show-option -Wno-unused-local-typedefs -Wno-attributes -Wno-psabi -Wno-error=unused-variable -DBOOST_DISABLE_ASSERTS -flto=auto -fipa-icf -flto-odr-type-merging -fno-fat-lto-objects -Wodr -fPIC -MMD -MF tmp/el8_amd64_gcc12/src/HeterogeneousCore/ROCmServices/test/testROCmService/test_main.cpp.d src/HeterogeneousCore/ROCmServices/test/test_main.cpp -o tmp/el8_amd64_gcc12/src/HeterogeneousCore/ROCmServices/test/testROCmService/test_main.cpp.o
In file included from src/HeterogeneousCore/ROCmServices/test/testROCmService.cpp:20:
poison/FWCore/Utilities/interface/ResourceInformation.h:1:2: error: #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
    1 | #error THIS FILE HAS BEEN REMOVED FROM THE PACKAGE.
      |  ^~~~~
src/HeterogeneousCore/ROCmServices/test/testROCmService.cpp: In function 'void ____C_A_T_C_H____T_E_S_T____0()':
src/HeterogeneousCore/ROCmServices/test/testROCmService.cpp:134:25: error: 'ResourceInformation' is not a member of 'edm'
  134 |       edm::Service<edm::ResourceInformation> ri;
</pre>

--- a/FWCore/AbstractServices/BuildFile.xml
+++ b/FWCore/AbstractServices/BuildFile.xml
@@ -1,0 +1,3 @@
+<export>
+  <lib name="1"/>
+</export>

--- a/FWCore/AbstractServices/interface/ResourceInformation.h
+++ b/FWCore/AbstractServices/interface/ResourceInformation.h
@@ -1,5 +1,5 @@
-#ifndef FWCore_Utilities_ResourceInformation_h
-#define FWCore_Utilities_ResourceInformation_h
+#ifndef FWCore_AbstractServices_ResourceInformation_h
+#define FWCore_AbstractServices_ResourceInformation_h
 
 /** \class edm::ResourceInformation
 

--- a/FWCore/AbstractServices/interface/ResourceInformation.h
+++ b/FWCore/AbstractServices/interface/ResourceInformation.h
@@ -22,9 +22,7 @@ namespace edm {
     ResourceInformation const& operator=(ResourceInformation const&) = delete;
     virtual ~ResourceInformation();
 
-    enum class AcceleratorType { GPU };
-
-    virtual std::vector<AcceleratorType> const& acceleratorTypes() const = 0;
+    virtual std::vector<std::string> const& selectedAccelerators() const = 0;
     virtual std::vector<std::string> const& cpuModels() const = 0;
     virtual std::vector<std::string> const& gpuModels() const = 0;
 
@@ -36,7 +34,7 @@ namespace edm {
     virtual std::string const& cpuModelsFormatted() const = 0;
     virtual double cpuAverageSpeed() const = 0;
 
-    virtual void initializeAcceleratorTypes(std::vector<std::string> const& selectedAccelerators) = 0;
+    virtual void setSelectedAccelerators(std::vector<std::string> const& selectedAccelerators) = 0;
     virtual void setCPUModels(std::vector<std::string> const&) = 0;
     virtual void setGPUModels(std::vector<std::string> const&) = 0;
 

--- a/FWCore/AbstractServices/src/ResourceInformation.cc
+++ b/FWCore/AbstractServices/src/ResourceInformation.cc
@@ -1,0 +1,6 @@
+#include "FWCore/AbstractServices/interface/ResourceInformation.h"
+
+namespace edm {
+  ResourceInformation::ResourceInformation() = default;
+  ResourceInformation::~ResourceInformation() = default;
+}  // namespace edm

--- a/FWCore/Framework/BuildFile.xml
+++ b/FWCore/Framework/BuildFile.xml
@@ -4,6 +4,7 @@
 <use name="tbb"/>
 <use name="DataFormats/Common"/>
 <use name="DataFormats/Provenance"/>
+<use name="FWCore/AbstractServices"/>
 <use name="FWCore/Common"/>
 <use name="FWCore/Concurrency"/>
 <use name="FWCore/MessageLogger"/>

--- a/FWCore/Framework/src/ScheduleItems.cc
+++ b/FWCore/Framework/src/ScheduleItems.cc
@@ -6,6 +6,7 @@
 #include "DataFormats/Provenance/interface/ThinnedAssociationsHelper.h"
 #include "DataFormats/Provenance/interface/SubProcessParentageHelper.h"
 #include "DataFormats/Provenance/interface/SelectedProducts.h"
+#include "FWCore/AbstractServices/interface/ResourceInformation.h"
 #include "FWCore/Common/interface/SubProcessBlockHelper.h"
 #include "FWCore/Framework/interface/ExceptionActions.h"
 #include "FWCore/Framework/src/CommonParams.h"
@@ -19,7 +20,6 @@
 #include "FWCore/ServiceRegistry/interface/ServiceRegistry.h"
 #include "FWCore/Utilities/interface/BranchType.h"
 #include "FWCore/Utilities/interface/GetPassID.h"
-#include "FWCore/Utilities/interface/ResourceInformation.h"
 #include "FWCore/Version/interface/GetReleaseVersion.h"
 
 #include <memory>

--- a/FWCore/Framework/src/ScheduleItems.cc
+++ b/FWCore/Framework/src/ScheduleItems.cc
@@ -123,7 +123,7 @@ namespace edm {
     if (resourceInformationService.isAvailable()) {
       auto const& selectedAccelerators =
           parameterSet.getUntrackedParameter<std::vector<std::string>>("@selected_accelerators");
-      resourceInformationService->initializeAcceleratorTypes(selectedAccelerators);
+      resourceInformationService->setSelectedAccelerators(selectedAccelerators);
     }
 
     act_table_ = std::make_unique<ExceptionToActionTable>(parameterSet);

--- a/FWCore/Services/BuildFile.xml
+++ b/FWCore/Services/BuildFile.xml
@@ -1,9 +1,10 @@
 <use name="SimDataFormats/RandomEngine"/>
+<use name="FWCore/AbstractServices"/>
+<use name="FWCore/Catalog"/>
 <use name="FWCore/MessageLogger"/>
 <use name="FWCore/ParameterSet"/>
 <use name="FWCore/ServiceRegistry"/>
 <use name="FWCore/Utilities"/>
-<use name="FWCore/Catalog"/>
 <use name="boost"/>
 <use name="tinyxml2"/>
 <use name="rootcore"/>

--- a/FWCore/Services/plugins/CPU.cc
+++ b/FWCore/Services/plugins/CPU.cc
@@ -8,6 +8,7 @@
 // Original Author:  Natalia Garcia
 // CPU.cc: v 1.0 2009/01/08 11:31:07
 
+#include "FWCore/AbstractServices/interface/ResourceInformation.h"
 #include "FWCore/MessageLogger/interface/JobReport.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 #include "FWCore/ServiceRegistry/interface/ActivityRegistry.h"
@@ -16,7 +17,6 @@
 #include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
 #include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
 #include "FWCore/Utilities/interface/CPUServiceBase.h"
-#include "FWCore/Utilities/interface/ResourceInformation.h"
 
 #include "cpu_features/cpu_features_macros.h"
 

--- a/FWCore/Services/plugins/CondorStatusUpdater.cc
+++ b/FWCore/Services/plugins/CondorStatusUpdater.cc
@@ -1,8 +1,8 @@
 
 #include "DataFormats/Provenance/interface/ModuleDescription.h"
 #include "DataFormats/Provenance/interface/ParameterSetID.h"
+#include "FWCore/AbstractServices/interface/ResourceInformation.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
-#include "FWCore/Utilities/interface/ResourceInformation.h"
 #include "FWCore/Utilities/interface/TimingServiceBase.h"
 #include "FWCore/ServiceRegistry/interface/Service.h"
 #include "FWCore/ServiceRegistry/interface/ProcessContext.h"

--- a/FWCore/Services/plugins/ResourceInformationService.cc
+++ b/FWCore/Services/plugins/ResourceInformationService.cc
@@ -31,7 +31,7 @@ namespace edm {
 
       static void fillDescriptions(ConfigurationDescriptions&);
 
-      std::vector<AcceleratorType> const& acceleratorTypes() const final;
+      std::vector<std::string> const& selectedAccelerators() const final;
       std::vector<std::string> const& cpuModels() const final;
       std::vector<std::string> const& gpuModels() const final;
 
@@ -43,7 +43,7 @@ namespace edm {
       std::string const& cpuModelsFormatted() const final;
       double cpuAverageSpeed() const final;
 
-      void initializeAcceleratorTypes(std::vector<std::string> const& selectedAccelerators) final;
+      void setSelectedAccelerators(std::vector<std::string> const& selectedAccelerators) final;
       void setCPUModels(std::vector<std::string> const&) final;
       void setGPUModels(std::vector<std::string> const&) final;
 
@@ -59,7 +59,7 @@ namespace edm {
     private:
       void throwIfLocked() const;
 
-      std::vector<AcceleratorType> acceleratorTypes_;
+      std::vector<std::string> selectedAccelerators_;
       std::vector<std::string> cpuModels_;
       std::vector<std::string> gpuModels_;
 
@@ -87,8 +87,8 @@ namespace edm {
       descriptions.add("ResourceInformationService", desc);
     }
 
-    std::vector<ResourceInformation::AcceleratorType> const& ResourceInformationService::acceleratorTypes() const {
-      return acceleratorTypes_;
+    std::vector<std::string> const& ResourceInformationService::selectedAccelerators() const {
+      return selectedAccelerators_;
     }
 
     std::vector<std::string> const& ResourceInformationService::cpuModels() const { return cpuModels_; }
@@ -105,15 +105,9 @@ namespace edm {
 
     double ResourceInformationService::cpuAverageSpeed() const { return cpuAverageSpeed_; }
 
-    void ResourceInformationService::initializeAcceleratorTypes(std::vector<std::string> const& selectedAccelerators) {
+    void ResourceInformationService::setSelectedAccelerators(std::vector<std::string> const& selectedAccelerators) {
       if (!locked_) {
-        for (auto const& selected : selectedAccelerators) {
-          // Test if the string begins with "gpu-"
-          if (selected.rfind("gpu-", 0) == 0) {
-            acceleratorTypes_.push_back(AcceleratorType::GPU);
-            break;
-          }
-        }
+        selectedAccelerators_ = selectedAccelerators;
         locked_ = true;
       }
     }
@@ -182,16 +176,12 @@ namespace edm {
           }
         }
 
-        LogAbsolute("ResourceInformation") << "    acceleratorTypes:";
-        if (acceleratorTypes().empty()) {
+        LogAbsolute("ResourceInformation") << "    selectedAccelerators:";
+        if (selectedAccelerators().empty()) {
           LogAbsolute("ResourceInformation") << "        None";
         } else {
-          for (auto const& iter : acceleratorTypes()) {
-            std::string acceleratorTypeString("unknown type");
-            if (iter == AcceleratorType::GPU) {
-              acceleratorTypeString = std::string("GPU");
-            }
-            LogAbsolute("ResourceInformation") << "        " << acceleratorTypeString;
+          for (auto const& iter : selectedAccelerators()) {
+            LogAbsolute("ResourceInformation") << "        " << iter;
           }
         }
         LogAbsolute("ResourceInformation") << "    nvidiaDriverVersion: " << nvidiaDriverVersion();

--- a/FWCore/Services/plugins/ResourceInformationService.cc
+++ b/FWCore/Services/plugins/ResourceInformationService.cc
@@ -11,13 +11,13 @@
 
 */
 
+#include "FWCore/AbstractServices/interface/ResourceInformation.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 #include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
 #include "FWCore/ServiceRegistry/interface/ActivityRegistry.h"
 #include "FWCore/Utilities/interface/EDMException.h"
-#include "FWCore/Utilities/interface/ResourceInformation.h"
 
 #include <string>
 #include <vector>

--- a/FWCore/Services/test/test_resourceInformationService.sh
+++ b/FWCore/Services/test/test_resourceInformationService.sh
@@ -4,7 +4,10 @@ function die { echo Failure $1: status $2 ; exit $2 ; }
 
 cmsRun ${SCRAM_TEST_PATH}/test_resourceInformationService_cfg.py &> test_resourceInformationService.log || die "cmsRun test_resourceInformationService_cfg.py" $?
 
-grep -A 1 "acceleratorTypes:" test_resourceInformationService.log | grep "GPU" || die "Check for existence of GPU acceleratorType" $?
+grep -A 4 "selectedAccelerators:" test_resourceInformationService.log | grep "cpu" || die "Check for existence of cpu in selectedAccelerators" $?
+grep -A 4 "selectedAccelerators:" test_resourceInformationService.log | grep "gpu-foo" || die "Check for existence of gpu-foo in selectedAccelerators" $?
+grep -A 4 "selectedAccelerators:" test_resourceInformationService.log | grep "test1" || die "Check for existence of test1 in selectedAccelerators" $?
+grep -A 4 "selectedAccelerators:" test_resourceInformationService.log | grep "test2" || die "Check for existence of test2 in selectedAccelerators" $?
 grep -A 1 "cpu models:" test_resourceInformationService.log | grep "None" && die "Check there is at least one model (not None)" 1
 
 exit 0

--- a/FWCore/Utilities/src/ResourceInformation.cc
+++ b/FWCore/Utilities/src/ResourceInformation.cc
@@ -1,4 +1,0 @@
-#include "FWCore/Utilities/interface/ResourceInformation.h"
-
-edm::ResourceInformation::ResourceInformation() {}
-edm::ResourceInformation::~ResourceInformation() {}

--- a/HeterogeneousCore/CUDAServices/plugins/BuildFile.xml
+++ b/HeterogeneousCore/CUDAServices/plugins/BuildFile.xml
@@ -3,6 +3,7 @@
   <use name="tbb"/>
   <use name="DataFormats/Common"/>
   <use name="DataFormats/Provenance"/>
+  <use name="FWCore/AbstractServices"/>
   <use name="FWCore/MessageLogger"/>
   <use name="FWCore/ParameterSet"/>
   <use name="FWCore/ServiceRegistry"/>

--- a/HeterogeneousCore/CUDAServices/plugins/CUDAService.cc
+++ b/HeterogeneousCore/CUDAServices/plugins/CUDAService.cc
@@ -11,12 +11,12 @@
 #include <cuda_runtime.h>
 #include <nvml.h>
 
+#include "FWCore/AbstractServices/interface/ResourceInformation.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 #include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
 #include "FWCore/ServiceRegistry/interface/Service.h"
-#include "FWCore/Utilities/interface/ResourceInformation.h"
 #include "FWCore/Utilities/interface/ReusableObjectHolder.h"
 #include "HeterogeneousCore/CUDAServices/interface/CUDAInterface.h"
 #include "HeterogeneousCore/CUDAUtilities/interface/EventCache.h"

--- a/HeterogeneousCore/CUDAServices/test/BuildFile.xml
+++ b/HeterogeneousCore/CUDAServices/test/BuildFile.xml
@@ -1,5 +1,6 @@
 <use name="catch2"/>
 <use name="fmt"/>
+<use name="FWCore/AbstractServices"/>
 <use name="FWCore/ParameterSet"/>
 <use name="FWCore/ParameterSetReader"/>
 <use name="FWCore/PluginManager"/>

--- a/HeterogeneousCore/CUDAServices/test/testCUDAService.cpp
+++ b/HeterogeneousCore/CUDAServices/test/testCUDAService.cpp
@@ -10,6 +10,7 @@
 
 #include <catch.hpp>
 
+#include "FWCore/AbstractServices/interface/ResourceInformation.h"
 #include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "FWCore/ParameterSetReader/interface/ParameterSetReader.h"
@@ -17,7 +18,6 @@
 #include "FWCore/ServiceRegistry/interface/ServiceRegistry.h"
 #include "FWCore/ServiceRegistry/interface/ServiceToken.h"
 #include "FWCore/Utilities/interface/Exception.h"
-#include "FWCore/Utilities/interface/ResourceInformation.h"
 #include "HeterogeneousCore/CUDAServices/interface/CUDAInterface.h"
 
 namespace {

--- a/HeterogeneousCore/ROCmServices/plugins/BuildFile.xml
+++ b/HeterogeneousCore/ROCmServices/plugins/BuildFile.xml
@@ -1,6 +1,7 @@
 <iftool name="rocm">
   <use name="rocm"/>
   <use name="DataFormats/Provenance"/>
+  <use name="FWCore/AbstractServices"/>
   <use name="FWCore/MessageLogger"/>
   <use name="FWCore/ParameterSet"/>
   <use name="FWCore/ServiceRegistry"/>

--- a/HeterogeneousCore/ROCmServices/plugins/ROCmService.cc
+++ b/HeterogeneousCore/ROCmServices/plugins/ROCmService.cc
@@ -14,12 +14,12 @@
 #endif  // HIP_VERSION_MAJOR
 #include <rocm_smi/rocm_smi.h>
 
+#include "FWCore/AbstractServices/interface/ResourceInformation.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 #include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
 #include "FWCore/ServiceRegistry/interface/Service.h"
-#include "FWCore/Utilities/interface/ResourceInformation.h"
 #include "HeterogeneousCore/ROCmServices/interface/ROCmInterface.h"
 #include "HeterogeneousCore/ROCmUtilities/interface/hipCheck.h"
 #include "HeterogeneousCore/ROCmUtilities/interface/rsmiCheck.h"

--- a/HeterogeneousCore/ROCmServices/test/BuildFile.xml
+++ b/HeterogeneousCore/ROCmServices/test/BuildFile.xml
@@ -1,5 +1,6 @@
 <use name="catch2"/>
 <use name="fmt"/>
+<use name="FWCore/AbstractServices"/>
 <use name="FWCore/ParameterSet"/>
 <use name="FWCore/ParameterSetReader"/>
 <use name="FWCore/PluginManager"/>

--- a/HeterogeneousCore/ROCmServices/test/testROCmService.cpp
+++ b/HeterogeneousCore/ROCmServices/test/testROCmService.cpp
@@ -10,6 +10,7 @@
 
 #include <catch.hpp>
 
+#include "FWCore/AbstractServices/interface/ResourceInformation.h"
 #include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "FWCore/ParameterSetReader/interface/ParameterSetReader.h"
@@ -17,7 +18,6 @@
 #include "FWCore/ServiceRegistry/interface/ServiceRegistry.h"
 #include "FWCore/ServiceRegistry/interface/ServiceToken.h"
 #include "FWCore/Utilities/interface/Exception.h"
-#include "FWCore/Utilities/interface/ResourceInformation.h"
 #include "HeterogeneousCore/ROCmServices/interface/ROCmInterface.h"
 
 namespace {

--- a/PhysicsTools/PyTorch/test/BuildFile.xml
+++ b/PhysicsTools/PyTorch/test/BuildFile.xml
@@ -15,6 +15,7 @@
     <use name="cppunit"/>
     <use name="cuda"/>
     <use name="pytorch"/>
+    <use name="FWCore/AbstractServices"/>
     <use name="FWCore/ParameterSet"/>
     <use name="FWCore/ParameterSetReader"/>
     <use name="FWCore/PluginManager"/>

--- a/PhysicsTools/PyTorch/test/testBaseCUDA.h
+++ b/PhysicsTools/PyTorch/test/testBaseCUDA.h
@@ -11,6 +11,7 @@
 #include <cppunit/extensions/HelperMacros.h>
 #include <stdexcept>
 
+#include "FWCore/AbstractServices/interface/ResourceInformation.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
 #include "FWCore/ParameterSetReader/interface/ParameterSetReader.h"
@@ -20,7 +21,6 @@
 #include "FWCore/ServiceRegistry/interface/ServiceRegistry.h"
 #include "FWCore/ServiceRegistry/interface/ServiceToken.h"
 #include "FWCore/Utilities/interface/Exception.h"
-#include "FWCore/Utilities/interface/ResourceInformation.h"
 
 class testBasePyTorchCUDA : public CppUnit::TestFixture {
 public:

--- a/PhysicsTools/TensorFlow/BuildFile.xml
+++ b/PhysicsTools/TensorFlow/BuildFile.xml
@@ -1,5 +1,6 @@
 <use name="tbb"/>
 <use name="tensorflow-cc"/>
+<use name="FWCore/AbstractServices"/>
 <use name="FWCore/Framework"/>
 <use name="FWCore/MessageLogger"/>
 <use name="FWCore/Utilities"/>

--- a/PhysicsTools/TensorFlow/src/TensorFlow.cc
+++ b/PhysicsTools/TensorFlow/src/TensorFlow.cc
@@ -6,9 +6,9 @@
  */
 
 #include "PhysicsTools/TensorFlow/interface/TensorFlow.h"
+#include "FWCore/AbstractServices/interface/ResourceInformation.h"
 #include "FWCore/MessageLogger/interface/MessageLogger.h"
 #include "FWCore/ServiceRegistry/interface/Service.h"
-#include "FWCore/Utilities/interface/ResourceInformation.h"
 
 namespace tensorflow {
 

--- a/PhysicsTools/TensorFlow/src/TensorFlow.cc
+++ b/PhysicsTools/TensorFlow/src/TensorFlow.cc
@@ -41,7 +41,7 @@ namespace tensorflow {
     }
     // NVidia GPU
     else if (backend == Backend::cuda) {
-      if (not ri->nvidiaDriverVersion().empty()) {
+      if (ri->hasGpuNvidia()) {
         // Check if one GPU device is visible to TF
         // If not, an exception is raised --> this can happen in case of driver version mismatch
         // or missing CUDA support in TF compilation
@@ -73,7 +73,7 @@ namespace tensorflow {
     // Get NVidia GPU if possible or fallback to CPU
     else if (backend == Backend::best) {
       // Check if a Nvidia GPU is availabl
-      if (not ri->nvidiaDriverVersion().empty()) {
+      if (ri->hasGpuNvidia()) {
         // Take only the first GPU in the CUDA_VISIBLE_DEVICE list
         (*_options.config.mutable_device_count())["GPU"] = 1;
         _options.config.mutable_gpu_options()->set_visible_device_list("0");

--- a/PhysicsTools/TensorFlow/test/BuildFile.xml
+++ b/PhysicsTools/TensorFlow/test/BuildFile.xml
@@ -12,6 +12,7 @@
     <use name="cppunit"/>
     <use name="cuda"/>
     <use name="tensorflow-cc"/>
+    <use name="FWCore/AbstractServices"/>
     <use name="FWCore/ParameterSet"/>
     <use name="FWCore/ParameterSetReader"/>
     <use name="FWCore/PluginManager"/>
@@ -26,6 +27,7 @@
 <bin name="testTFMetaGraphLoading" file="testRunner.cpp,testMetaGraphLoading.cc">
   <use name="boost_filesystem"/>
   <use name="cppunit"/>
+  <use name="FWCore/AbstractServices"/>
   <use name="PhysicsTools/TensorFlow"/>
 </bin>
 
@@ -35,6 +37,7 @@
     <use name="catch2"/>
     <use name="cppunit"/>
     <use name="cuda"/>
+    <use name="FWCore/AbstractServices"/>
     <use name="FWCore/ParameterSet"/>
     <use name="FWCore/ParameterSetReader"/>
     <use name="FWCore/PluginManager"/>
@@ -82,6 +85,7 @@
     <use name="catch2"/>
     <use name="cppunit"/>
     <use name="cuda"/>
+    <use name="FWCore/AbstractServices"/>
     <use name="FWCore/ParameterSet"/>
     <use name="FWCore/ParameterSetReader"/>
     <use name="FWCore/PluginManager"/>
@@ -105,6 +109,7 @@
     <use name="catch2"/>
     <use name="cppunit"/>
     <use name="cuda"/>
+    <use name="FWCore/AbstractServices"/>
     <use name="FWCore/ParameterSet"/>
     <use name="FWCore/ParameterSetReader"/>
     <use name="FWCore/PluginManager"/>
@@ -129,6 +134,7 @@
     <use name="catch2"/>
     <use name="cppunit"/>
     <use name="cuda"/>
+    <use name="FWCore/AbstractServices"/>
     <use name="FWCore/ParameterSet"/>
     <use name="FWCore/ParameterSetReader"/>
     <use name="FWCore/PluginManager"/>
@@ -160,6 +166,7 @@
     <use name="catch2"/>
     <use name="cppunit"/>
     <use name="cuda"/>
+    <use name="FWCore/AbstractServices"/>
     <use name="FWCore/ParameterSet"/>
     <use name="FWCore/ParameterSetReader"/>
     <use name="FWCore/PluginManager"/>

--- a/PhysicsTools/TensorFlow/test/testBaseCUDA.h
+++ b/PhysicsTools/TensorFlow/test/testBaseCUDA.h
@@ -13,6 +13,7 @@
 #include <stdexcept>
 #include "catch.hpp"
 
+#include "FWCore/AbstractServices/interface/ResourceInformation.h"
 #include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
 #include "FWCore/ParameterSetReader/interface/ParameterSetReader.h"
@@ -22,7 +23,6 @@
 #include "FWCore/ServiceRegistry/interface/ServiceRegistry.h"
 #include "FWCore/ServiceRegistry/interface/ServiceToken.h"
 #include "FWCore/Utilities/interface/Exception.h"
-#include "FWCore/Utilities/interface/ResourceInformation.h"
 #include "HeterogeneousCore/CUDAServices/interface/CUDAInterface.h"
 #include "HeterogeneousCore/CUDAUtilities/interface/requireDevices.h"
 

